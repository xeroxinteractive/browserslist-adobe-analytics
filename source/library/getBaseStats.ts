import { agents } from 'caniuse-lite';
import { Stats } from 'browserslist';

/**
 * Returns browserslist statistics data with all the %'s set to 0.
 *
 * @remarks
 * Uses caniuse-lite data get all the browsers/versions.
 * Inspired by https://github.com/browserslist/browserslist-ga/blob/89a0c2dbf173632938cb1a8de24f9c3a0f4dc876/src/caniuse-agent-data.js
 * @returns Browserslist statistics.
 */
export default function getBaseStats(): Stats {
  const baseStats: Stats = {};
  const out = Object.entries(agents).reduce(
    (stats, [browser, agent]): Stats => {
      if (agent) {
        stats[browser] = Object.entries(agent.release_date)
          .sort((a, b) => {
            if (a[1] === undefined) {
              return -1;
            } else if (b[1] === undefined) {
              return 1;
            } else {
              return b[1] - a[1];
            }
          })
          .reduce((versions, [version]) => {
            versions[version] = 0;
            return versions;
          }, {} as { [version: string]: number });
      }
      return stats;
    },
    baseStats
  );
  return out;
}
