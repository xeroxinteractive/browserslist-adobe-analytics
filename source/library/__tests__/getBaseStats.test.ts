import getBaseStats from '../getBaseStats';

test('expected data', () => {
  expect(getBaseStats()).toMatchSnapshot();
});

test('version order', () => {
  const stats = getBaseStats();
  for (const [browser, versions] of Object.entries(stats)) {
    expect(Object.keys(versions)).toMatchSnapshot(browser);
  }
});
