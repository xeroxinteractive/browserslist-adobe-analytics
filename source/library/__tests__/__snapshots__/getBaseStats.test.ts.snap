// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actual data 1`] = `
Object {
  "and_chr": Object {
    "80": 0,
  },
  "and_ff": Object {
    "68": 0,
  },
  "and_qq": Object {
    "1.2": 0,
  },
  "and_uc": Object {
    "12.12": 0,
  },
  "android": Object {
    "2.1": 0,
    "2.2": 0,
    "2.3": 0,
    "3": 0,
    "4": 0,
    "4.1": 0,
    "4.2-4.3": 0,
    "4.4": 0,
    "4.4.3-4.4.4": 0,
    "80": 0,
  },
  "baidu": Object {
    "7.12": 0,
  },
  "bb": Object {
    "10": 0,
    "7": 0,
  },
  "chrome": Object {
    "10": 0,
    "11": 0,
    "12": 0,
    "13": 0,
    "14": 0,
    "15": 0,
    "16": 0,
    "17": 0,
    "18": 0,
    "19": 0,
    "20": 0,
    "21": 0,
    "22": 0,
    "23": 0,
    "24": 0,
    "25": 0,
    "26": 0,
    "27": 0,
    "28": 0,
    "29": 0,
    "30": 0,
    "31": 0,
    "32": 0,
    "33": 0,
    "34": 0,
    "35": 0,
    "36": 0,
    "37": 0,
    "38": 0,
    "39": 0,
    "4": 0,
    "40": 0,
    "41": 0,
    "42": 0,
    "43": 0,
    "44": 0,
    "45": 0,
    "46": 0,
    "47": 0,
    "48": 0,
    "49": 0,
    "5": 0,
    "50": 0,
    "51": 0,
    "52": 0,
    "53": 0,
    "54": 0,
    "55": 0,
    "56": 0,
    "57": 0,
    "58": 0,
    "59": 0,
    "6": 0,
    "60": 0,
    "61": 0,
    "62": 0,
    "63": 0,
    "64": 0,
    "65": 0,
    "66": 0,
    "67": 0,
    "68": 0,
    "69": 0,
    "7": 0,
    "70": 0,
    "71": 0,
    "72": 0,
    "73": 0,
    "74": 0,
    "75": 0,
    "76": 0,
    "77": 0,
    "78": 0,
    "79": 0,
    "8": 0,
    "80": 0,
    "81": 0,
    "82": 0,
    "83": 0,
    "9": 0,
  },
  "edge": Object {
    "12": 0,
    "13": 0,
    "14": 0,
    "15": 0,
    "16": 0,
    "17": 0,
    "18": 0,
    "79": 0,
    "80": 0,
  },
  "firefox": Object {
    "10": 0,
    "11": 0,
    "12": 0,
    "13": 0,
    "14": 0,
    "15": 0,
    "16": 0,
    "17": 0,
    "18": 0,
    "19": 0,
    "2": 0,
    "20": 0,
    "21": 0,
    "22": 0,
    "23": 0,
    "24": 0,
    "25": 0,
    "26": 0,
    "27": 0,
    "28": 0,
    "29": 0,
    "3": 0,
    "3.5": 0,
    "3.6": 0,
    "30": 0,
    "31": 0,
    "32": 0,
    "33": 0,
    "34": 0,
    "35": 0,
    "36": 0,
    "37": 0,
    "38": 0,
    "39": 0,
    "4": 0,
    "40": 0,
    "41": 0,
    "42": 0,
    "43": 0,
    "44": 0,
    "45": 0,
    "46": 0,
    "47": 0,
    "48": 0,
    "49": 0,
    "5": 0,
    "50": 0,
    "51": 0,
    "52": 0,
    "53": 0,
    "54": 0,
    "55": 0,
    "56": 0,
    "57": 0,
    "58": 0,
    "59": 0,
    "6": 0,
    "60": 0,
    "61": 0,
    "62": 0,
    "63": 0,
    "64": 0,
    "65": 0,
    "66": 0,
    "67": 0,
    "68": 0,
    "69": 0,
    "7": 0,
    "70": 0,
    "71": 0,
    "72": 0,
    "73": 0,
    "74": 0,
    "75": 0,
    "8": 0,
    "9": 0,
  },
  "ie": Object {
    "10": 0,
    "11": 0,
    "5.5": 0,
    "6": 0,
    "7": 0,
    "8": 0,
    "9": 0,
  },
  "ie_mob": Object {
    "10": 0,
    "11": 0,
  },
  "ios_saf": Object {
    "10.0-10.2": 0,
    "10.3": 0,
    "11.0-11.2": 0,
    "11.3-11.4": 0,
    "12.0-12.1": 0,
    "12.2-12.4": 0,
    "13.0-13.1": 0,
    "13.2": 0,
    "13.3": 0,
    "3.2": 0,
    "4.0-4.1": 0,
    "4.2-4.3": 0,
    "5.0-5.1": 0,
    "6.0-6.1": 0,
    "7.0-7.1": 0,
    "8": 0,
    "8.1-8.4": 0,
    "9.0-9.2": 0,
    "9.3": 0,
  },
  "kaios": Object {
    "2.5": 0,
  },
  "op_mini": Object {
    "all": 0,
  },
  "op_mob": Object {
    "10": 0,
    "11": 0,
    "11.1": 0,
    "11.5": 0,
    "12": 0,
    "12.1": 0,
    "46": 0,
  },
  "opera": Object {
    "10.0-10.1": 0,
    "10.5": 0,
    "10.6": 0,
    "11": 0,
    "11.1": 0,
    "11.5": 0,
    "11.6": 0,
    "12": 0,
    "12.1": 0,
    "15": 0,
    "16": 0,
    "17": 0,
    "18": 0,
    "19": 0,
    "20": 0,
    "21": 0,
    "22": 0,
    "23": 0,
    "24": 0,
    "25": 0,
    "26": 0,
    "27": 0,
    "28": 0,
    "29": 0,
    "30": 0,
    "31": 0,
    "32": 0,
    "33": 0,
    "34": 0,
    "35": 0,
    "36": 0,
    "37": 0,
    "38": 0,
    "39": 0,
    "40": 0,
    "41": 0,
    "42": 0,
    "43": 0,
    "44": 0,
    "45": 0,
    "46": 0,
    "47": 0,
    "48": 0,
    "49": 0,
    "50": 0,
    "51": 0,
    "52": 0,
    "53": 0,
    "54": 0,
    "55": 0,
    "56": 0,
    "57": 0,
    "58": 0,
    "60": 0,
    "62": 0,
    "63": 0,
    "64": 0,
    "65": 0,
    "66": 0,
    "9": 0,
    "9.5-9.6": 0,
  },
  "safari": Object {
    "10": 0,
    "10.1": 0,
    "11": 0,
    "11.1": 0,
    "12": 0,
    "12.1": 0,
    "13": 0,
    "3.1": 0,
    "3.2": 0,
    "4": 0,
    "5": 0,
    "5.1": 0,
    "6": 0,
    "6.1": 0,
    "7": 0,
    "7.1": 0,
    "8": 0,
    "9": 0,
    "9.1": 0,
    "TP": 0,
  },
  "samsung": Object {
    "10.1": 0,
    "4": 0,
    "5.0-5.4": 0,
    "6.2-6.4": 0,
    "7.2-7.4": 0,
    "8.2": 0,
    "9.2": 0,
  },
}
`;

exports[`actual ordering: and_chr 1`] = `
Array [
  "80",
]
`;

exports[`actual ordering: and_ff 1`] = `
Array [
  "68",
]
`;

exports[`actual ordering: and_qq 1`] = `
Array [
  "1.2",
]
`;

exports[`actual ordering: and_uc 1`] = `
Array [
  "12.12",
]
`;

exports[`actual ordering: android 1`] = `
Array [
  "3",
  "4",
  "80",
  "4.4.3-4.4.4",
  "4.4",
  "4.2-4.3",
  "4.1",
  "2.3",
  "2.2",
  "2.1",
]
`;

exports[`actual ordering: baidu 1`] = `
Array [
  "7.12",
]
`;

exports[`actual ordering: bb 1`] = `
Array [
  "7",
  "10",
]
`;

exports[`actual ordering: chrome 1`] = `
Array [
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "12",
  "13",
  "14",
  "15",
  "16",
  "17",
  "18",
  "19",
  "20",
  "21",
  "22",
  "23",
  "24",
  "25",
  "26",
  "27",
  "28",
  "29",
  "30",
  "31",
  "32",
  "33",
  "34",
  "35",
  "36",
  "37",
  "38",
  "39",
  "40",
  "41",
  "42",
  "43",
  "44",
  "45",
  "46",
  "47",
  "48",
  "49",
  "50",
  "51",
  "52",
  "53",
  "54",
  "55",
  "56",
  "57",
  "58",
  "59",
  "60",
  "61",
  "62",
  "63",
  "64",
  "65",
  "66",
  "67",
  "68",
  "69",
  "70",
  "71",
  "72",
  "73",
  "74",
  "75",
  "76",
  "77",
  "78",
  "79",
  "80",
  "81",
  "82",
  "83",
]
`;

exports[`actual ordering: edge 1`] = `
Array [
  "12",
  "13",
  "14",
  "15",
  "16",
  "17",
  "18",
  "79",
  "80",
]
`;

exports[`actual ordering: firefox 1`] = `
Array [
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "12",
  "13",
  "14",
  "15",
  "16",
  "17",
  "18",
  "19",
  "20",
  "21",
  "22",
  "23",
  "24",
  "25",
  "26",
  "27",
  "28",
  "29",
  "30",
  "31",
  "32",
  "33",
  "34",
  "35",
  "36",
  "37",
  "38",
  "39",
  "40",
  "41",
  "42",
  "43",
  "44",
  "45",
  "46",
  "47",
  "48",
  "49",
  "50",
  "51",
  "52",
  "53",
  "54",
  "55",
  "56",
  "57",
  "58",
  "59",
  "60",
  "61",
  "62",
  "63",
  "64",
  "65",
  "66",
  "67",
  "68",
  "69",
  "70",
  "71",
  "72",
  "73",
  "74",
  "75",
  "3.6",
  "3.5",
]
`;

exports[`actual ordering: ie 1`] = `
Array [
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "5.5",
]
`;

exports[`actual ordering: ie_mob 1`] = `
Array [
  "10",
  "11",
]
`;

exports[`actual ordering: ios_saf 1`] = `
Array [
  "8",
  "13.2",
  "13.0-13.1",
  "12.2-12.4",
  "12.0-12.1",
  "11.3-11.4",
  "11.0-11.2",
  "10.3",
  "10.0-10.2",
  "9.3",
  "9.0-9.2",
  "8.1-8.4",
  "7.0-7.1",
  "6.0-6.1",
  "5.0-5.1",
  "4.2-4.3",
  "4.0-4.1",
  "3.2",
  "13.3",
]
`;

exports[`actual ordering: kaios 1`] = `
Array [
  "2.5",
]
`;

exports[`actual ordering: op_mini 1`] = `
Array [
  "all",
]
`;

exports[`actual ordering: op_mob 1`] = `
Array [
  "10",
  "11",
  "12",
  "46",
  "12.1",
  "11.5",
  "11.1",
]
`;

exports[`actual ordering: opera 1`] = `
Array [
  "9",
  "11",
  "12",
  "15",
  "16",
  "17",
  "18",
  "19",
  "20",
  "21",
  "22",
  "23",
  "24",
  "25",
  "26",
  "27",
  "28",
  "29",
  "30",
  "31",
  "32",
  "33",
  "34",
  "35",
  "36",
  "37",
  "38",
  "39",
  "40",
  "41",
  "42",
  "43",
  "44",
  "45",
  "46",
  "47",
  "48",
  "49",
  "50",
  "51",
  "52",
  "53",
  "54",
  "55",
  "56",
  "57",
  "58",
  "60",
  "62",
  "63",
  "64",
  "65",
  "66",
  "12.1",
  "11.6",
  "11.5",
  "11.1",
  "10.6",
  "10.5",
  "10.0-10.1",
  "9.5-9.6",
]
`;

exports[`actual ordering: safari 1`] = `
Array [
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "12",
  "13",
  "12.1",
  "11.1",
  "10.1",
  "9.1",
  "7.1",
  "6.1",
  "5.1",
  "3.2",
  "3.1",
  "TP",
]
`;

exports[`actual ordering: samsung 1`] = `
Array [
  "4",
  "10.1",
  "9.2",
  "8.2",
  "7.2-7.4",
  "6.2-6.4",
  "5.0-5.4",
]
`;
